rag-openai-qdrant/rag-openai-qdrant.ipynb

```diff
@@ -0,0 +1,672 @@
+      ]
+     },
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "context = \"\\n\".join(r.document for r in results)\n",
+    "context"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a2c04a4e",
+   "metadata": {},
+   "source": [
+    "Finally, let's build a metaprompt that will force our LLM to use the provided context. That effectively converts the knowledge-oriented task into a language task and hopefully reduces the chances of hallucinations. It also should make the response sound more relevant."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "1fc9a98b",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n",
+      "You are a software architect. \n",
+      "Answer the following question using the provided context. \n",
+      "If you can't find the answer, do not pretend you know it, but answer \"I don't know\".\n",
+      "\n",
+      "Question: What tools should I need to use to build a web service using vector embeddings for search?\n",
+      "\n",
+      "Context: \n",
+      "Qdrant is a vector database & vector similarity search engine. It deploys as an API service providing search for the nearest high-dimensional vectors. With Qdrant, embeddings or neural network encoders can be turned into full-fledged applications for matching, searching, recommending, and much more!\n",
+      "FastAPI is a modern, fast (high-performance), web framework for building APIs with Python 3.7+ based on standard Python type hints.\n",
+      "SentenceTransformers is a Python framework for state-of-the-art sentence, text and image embeddings. You can use this framework to compute sentence / text embeddings for more than 100 languages. These embeddings can then be compared e.g. with cosine-similarity to find sentences with a similar meaning. This can be useful for semantic textual similar, semantic search, or paraphrase mining.\n",
+      "\n",
+      "Answer:\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "metaprompt = f\"\"\"\n",
+    "You are a software architect. \n",
+    "Answer the following question using the provided context. \n",
+    "If you can't find the answer, do not pretend you know it, but answer \"I don't know\".\n",
+    "\n",
+    "Question: {prompt.strip()}\n",
+    "\n",
+    "Context: \n",
+    "{context.strip()}\n",
+    "\n",
+    "Answer:\n",
+    "\"\"\"\n",
```

Comment:
```suggestion
    "\"\"\"\n",
    "\"\"\"\n",
```

Trying to make the code look a bit more clear. You could even put a comment above the print statement to help 

#look at the full metaprompt